modules/reactor/src/meshgenerators/ExtraElementIDConnectionFixup.C

```diff
@@ -0,0 +1,231 @@
+  params.addClassDescription(
+      "Check and change extra element IDs so that all fine elements with the same "
+      "element ID are connected through sides.");
+  return params;
+}
+
+ExtraElementIDConnectionFixup::ExtraElementIDConnectionFixup(const InputParameters & params)
+  : MeshGenerator(params), _input(getMesh("input")), _verbose(getParam<bool>("verbose"))
+{
+}
+
+std::unique_ptr<MeshBase>
+ExtraElementIDConnectionFixup::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  auto src_id_name = getParam<std::string>("extra_element_id");
+  if (!mesh->has_elem_integer(src_id_name))
+    mooseError("The source element ID does not exist on the input mesh");
+
+  // need to serialize the mesh for connection fixup
+  // Note: When replicated mesh is used or distributed mesh has not been distributed through
+  //       prepare_for_use, this is a null operation.
+  MeshSerializer tm(*mesh);
+
+  auto src_id = mesh->get_elem_integer_index(src_id_name);
+
+  // a map from coarse element ids to fine element ids
+  std::map<dof_id_type, std::set<dof_id_type>> map_coarse_to_fine;
+  // a map from nodes to connected fine element ids
+  std::map<dof_id_type, std::set<dof_id_type>> map_node_to_fine;
+  // a set for coarse eleemnt IDs
+  std::set<dof_id_type> coarse_ids;
+
+  // build the two maps and one sets above by doing one element loop
+  for (auto & elem : mesh->element_ptr_range())
+  {
+    dof_id_type id = elem->get_extra_integer(src_id);
+    if (id == DofObject::invalid_id)
+      continue;
+
+    map_coarse_to_fine[id].insert(elem->id());
+    coarse_ids.insert(id);
+
+    for (auto & node : elem->node_ref_range())
+      map_node_to_fine[node.id()].insert(elem->id());
+  }
+
+  if (_verbose)
+    _console << " Number of coarse elements before fixup: " << coarse_ids.size() << std::endl;
+
+  while (!coarse_ids.empty())
+  {
+    const dof_id_type coarse_id = *coarse_ids.begin();
+    coarse_ids.erase(coarse_id);
+    const auto & fine_ids = map_coarse_to_fine[coarse_id];
+
+    // the coarse element could contain no fine element due to fixup
+    if (fine_ids.size() == 0)
+      continue;
+
+    // merge the coarse element with only a single fine element into its neighboring coarse element
+    if (fine_ids.size() == 1)
+    {
+      Elem * elem = mesh->elem_ptr(*fine_ids.begin());
+      for (auto neig : elem->neighbor_ptr_range())
+        if (neig)
+        {
+          const dof_id_type newid = neig->get_extra_integer(src_id);
+          if (_verbose)
+            _console << " Single element with ID " << elem->id() << ": " << coarse_id << " => "
+                     << newid << std::endl;
+          elem->set_extra_integer(src_id, newid);
+          map_coarse_to_fine[coarse_id].erase(elem->id());
+          map_coarse_to_fine[newid].insert(elem->id());
+
+          // put changed ids into re-consideration
+          coarse_ids.insert(newid);
+          break;
+        }
+
+      continue;
+    }
```

Comment:
When we overlay a regular grid on an unstructured grid, we could have one regular grid just cutting a small tiny piece of the unstructured grid on its boundary with only one element. This is the reason we want to re-assign the id for this one element. This assumes that a coarse element has more than one fine element. Yes, I think this can run after the fixup. I will move it out of this loop and add a parameter to control if we want this to happen or not.